streaming/src/test/scala/org/apache/spark/streaming/ReceiverSuite.scala

```diff
@@ -212,6 +212,86 @@ class ReceiverSuite extends TestSuiteBase with Timeouts with Serializable {
     )
   }
 
+  // helps test congestion strategies with a constant rate per Block
+  // verifies the rate of messages produced by one of the congestion strategies
+  def testBlockGeneratorCongestion(congestionStrategy: CongestionStrategy,
+                                   blockIntervalMs: Int,
+                                   maxRatePerBlock: Int,
+                                   expectedBlocks: Int,
+                                   errorBound: Double = 0.05): Unit = {
+    congestionStrategy.onBlockBoundUpdate(maxRatePerBlock)
+    val blockGeneratorListener = new FakeBlockGeneratorListenerWithStrategy(congestionStrategy)
+    val conf = new SparkConf().set("spark.streaming.blockInterval", s"${blockIntervalMs}ms")
+    val blockGenerator = new BlockGenerator(blockGeneratorListener, 1, conf)
+
+    val waitTime = expectedBlocks * blockIntervalMs
+    val expectedMessages = maxRatePerBlock * expectedBlocks
+    val expectedMessagesPerBlock = maxRatePerBlock
+    val generatedData = new ArrayBuffer[Int]
+
+    // Generate blocks
+    val startTime = System.currentTimeMillis()
+    blockGenerator.start()
+    var count = 0
+    while(System.currentTimeMillis - startTime < waitTime) {
```

Comment:
What should be the acceptable skew there ? The idea of this call to `System.currentTimeMillis` is to keep generating Ints for at least an `expectedBlocks` number of blocks. To achieve  that, you need to generate for at least (`expectedBlocks`-1) x `blockInterval` + ε. In particular, that ε amount of time should generate more than 1000 integers (to test that the bound is respected in the last block). Experimentally, we need at (the very) most 1ms on a single thread to generate those 1000 integers (this is not RNG, just incremented Ints).

We set `waitTime`, the duration of generation to be `expectedBlocks` x `blockInterval`. We have, really, nearly one entire block interval more than what's needed. So the acceptable skew needs to be between an order of magnitude lower than `blockInterval` - ε ≅ `blockInterval` - 1ms.

With Spark Streaming, you can expect minimal batch sizes to be ½ second. Block sizes are about (batch interval / sizeof(cluster)), so on the order of 1/200 second. Here, in particular, we test with 100ms. So skews of up to 9ms are ok.

I think the call to `currentTimeMillis` is about correct here, and hence that we don't need to break out `nanoTime`. Or did I miss something ?